### ISSUE TITLE
CanvasItem constructor missing but required (C#)

### ISSUE THREAD

**Issue by gusg21** [godotengine/godot#55549]
### Godot version

3.4.stable

### System information

macOS Big Sur, 2017 iMac, i7

### Issue description

When trying to extend a CanvasItem in C#, Godot throws an error saying that ```/Users/.../ShowInGame.cs(4,14): 'CanvasItem' does not contain a constructor that takes 0 arguments```. However, the C# metadata for CanvasItem does not contain a constructor and I can find no documentation for it on the internet. Is CanvasItem not intended to be extended? If so, why expose it as an extendable class?

Here is my code trying to extend CanvasItem:
```csharp
using Godot;
using System;

public class ShowInGame : CanvasItem
{
    public override void _Ready()
    {
        Visible = true;

        base._Ready();
    }
}
```

### Steps to reproduce

If you try to compile, run, or build the project, you can see the error described above.

### Minimal reproduction project

[MinCanvasItemError.zip](https://github.com/godotengine/godot/files/7643206/MinCanvasItemError.zip)

**Comment by raulsntos** [comment]
`CanvasItem` is a non-instantiable class, therefore the [`bindings_generator.cpp`](https://github.com/godotengine/godot/blob/892a5a72cd9b25dc660671b9f4244fd842884b9f/modules/mono/editor/bindings_generator.cpp#L1435-L1438) generates an internal constructor when generating the C# class. Since the constructor is internal your class can't call it.

The reason why it's exposed as a public class is that some public methods may use it as the type of their parameters or the return type, so you need access to it.

---

Duplicate of #35482

**Comment by kleonc** [comment]
> `CanvasItem` is a non-instantiable class, therefore the [`bindings_generator.cpp`](https://github.com/godotengine/godot/blob/892a5a72cd9b25dc660671b9f4244fd842884b9f/modules/mono/editor/bindings_generator.cpp#L1435-L1438) generates an internal constructor when generating the C# class.

@raulsntos Seems like each non-instantiable Godot class (`CanvasItem`, `BaseButton`, `CollisionObject`, etc.) [is being generated as abstract](https://github.com/godotengine/godot/blob/892a5a72cd9b25dc660671b9f4244fd842884b9f/modules/mono/editor/bindings_generator.cpp#L1297) so it can't be instantiated on its own no matter what access modifier its constructor has. So generating the constructors as internal isn't needed to prevent instantiating these classes. And indeed constructors being internal prevent the users extending these classes. So the question is: is extending disallowed intentionally for these classes? If not and extending should be allowed then it could be done by making constructors be generated as protected.

**Comment by raulsntos** [comment]
Personally I think extending should be allowed, I think it works in GDScript so it should work in C# too.

Unfortunately, I think the `abstract` keyword will be removed in the future (4.1?) because with the switch from Mono to CLR these classes can no longer be abstract (they are instantiated by the editor, Mono allowed us to instantiate abstract classes but I believe that's not possible with the CLR).

Therefore, without the `abstract` keyword the only way to prevent users from instantiating these classes is to hide the constructor (that's why it's internal, it can be instantiated by the editor but not the user).

**Comment by cgbeutler** [comment]
Just found this when trying to convert my custom button class from GDScript to CSharp.
The documentation for `BaseButton` says it can/should be derived from. You can derive from all of them on the GDScript side and it works as expected. This is definitely a bug.

The default constructors should probably be made `protected` instead of `internal`. That would allow them to be extended, (regardless of what happens with abstract.) Protected constructors can still be called using reflection, so I'm _guessing_ it would be doable in CLR?

**Comment by raulsntos** [comment]
> The default constructors should probably be made `protected` instead of `internal`. That would allow them to be extended, (regardless of what happens with abstract.) Protected constructors can still be called using reflection, so I'm _guessing_ it would be doable in CLR?

That's what I was thinking, or better yet [`protected internal`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/protected-internal) which is essentially `protected` for external assemblies (the game project) and `public` for the defining assembly (GodotSharp.dll) and [friends](https://docs.microsoft.com/en-us/dotnet/standard/assembly/friend) which would avoid the need for using reflection.

However, I'm not sure if there are other reasons to not make the constructor accessible to users.

**Comment by cgbeutler** [comment]
> However, I'm not sure if there are other reasons to not make the constructor accessible to users.

Yeah, they could very well be public all the way. There is a _slight_ chance that someone may want to make one of them, then set the script later. Not sure why you'd do that, but it is technically how the GDScript side works.

**Comment by cgbeutler** [comment]
May be worth changing the title of this issue, as this seems to be a broader issue. Maybe something close to what kleonc said, "C# Abstract Godot types missing constructors" or something.

I have run into this issue again with `Texture` and `StyleBox`. The `StyleBox` one was especially annoying, as `Theme`s _require_ it to be derived from `StyleBox`. Had to hack around themes entirely to get what I wanted.

**Comment by akien-mga** [comment]
Related to #35484.

**Comment by akien-mga** [comment]
> I have run into this issue again with `Texture` and `StyleBox`. The `StyleBox` one was especially annoying, as `Theme`s _require_ it to be derived from `StyleBox`. Had to abandon themes to get what I wanted.

You can still use any type derived from `StyleBox` or `Texture`. E.g. `StyleBoxFlat`.

This issue should only be a problem for users who want to implement their own custom `Texture` or `StyleBox` type without basing it on an existing instantiable type. We're working on fixing this in 4.0.

**Comment by neikeq** [comment]
Reading the code, I doubt you can inherit CanvasItem in GDScript. You should be getting the error: `Class type: 'CanvasItem' is not ininstantiable`.

**Comment by cgbeutler** [comment]
@neikeq In GDScript you can inherit from CanvasItem and then attach that script to a derived type like Control or Node2D. Never tried base-class scripts in C# myself, so I don't know were that stands, atm.

**Comment by neikeq** [comment]
> In GDScript you can inherit from CanvasItem and then attach that script to a derived type like Control or Node2D

That's something we definitely won't support in C#. It's a very gray area, and it won't even be technically possible if we move to GDExtension in the future.

**Comment by neikeq** [comment]
Since this issue applies to all languages and not only C#, we should either close it or re-word/re-open it as a global Godot issue.

**Comment by cgbeutler** [comment]
> Since this issue applies to all languages and not only C#, we should either close it or re-word/re-open it as a global Godot issue.

Yeah, as this issue's new title indicates, this issue has morphed away from its discussion and `CanvasItem` and has become logistics on which abstract types like `BaseButton`, `CollisionObject`, `Texture`, `StyleBox`, etc. need to have accessible constructors in C#. We have also discussed whether those constructors could be public or protected or something else. At the heart of it all, this all has to do with the same bit of C# code generation, so I don't think it is an "all languages" thing.

Unless, of course, the @gusg21 feels like we have co-opted this issue and made it something else. If that's the case, I would be happy to open a new issue with a description to match the above paragraph and this issue can become a discussion on whether all languages should have a `CanvasItem` constructor or not.

**Comment by neikeq** [comment]
I'm speaking of CanvasItem for simplicity here, but it applies to all abstract classes. This problem exists for all languages because the original C++ class itself cannot be instantiated. Making the CanvasItem constructor public in C# won't change that; it would simply make a C# class pretend to be CanvasItem, when the actual C++ type is Node (what CanvasItem inherits, because that's the closest base class that can be instantiated).

You can't inherit **and** instantiate it in GDScript, either: calling `.new()` will result in an error. The reason it works when **attaching** to a CanvasItem derived type is a GDScript special thing, which wouldn't translate well to languages with bindings (I even wonder if this could have side effects in GDScript).

I'm not trying to co-opt this issue. I'm saying it cannot be fixed in C#. If we want to fix it, it has to be done globally, at the core.

**Comment by akien-mga** [comment]
See the core issue I linked above: https://github.com/godotengine/godot/issues/55549#issuecomment-1061855184

**Comment by 4CoDev** [comment]
I hope that VisualInstance will also be among those allowed to inherit

**Comment by FyiurAmron** [comment]
@akien-mga what's the current status here? I'm seeing this problem reported in different forms and different situations from time to time, but I can't really grasp what's the core dev team's stance on this - is it likely this will be resolved, or are workarounds (like using composition pattern etc.) the only hope for the future?
